mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,307 @@
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Types.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Support/Debug.h"
+
+#include <memory>
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+#define GEN_PASS_DEF_SPLITMULTIPLETAPESPASS
+#include "Catalyst/Transforms/Passes.h.inc"
+
+struct SplitMultipleTapesPass : public impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass> {
+    using impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass>::SplitMultipleTapesPassBase;
+
+    unsigned int countTapes(const func::FuncOp &func)
+    {
+        // Count the number of quantum.device operations in a function
+        unsigned int count = 0;
+        func->walk([&](Operation *op) {
+            if (op->getName().getStringRef() == "quantum.device") {
+                count++;
+            }
+        });
+        return count;
+    } // countTapes()
+
+    void collectOperationsForEachTape(
+        const func::FuncOp &func,
+        SmallVector<std::shared_ptr<SmallVector<Operation *>>> &OpsEachTape)
+    {
+        // During tracing, each tape starts with a qdevice_p primitive
+        // This means each tape starts with a quantum.device
+        // and ends with a quantum.device_release
+
+        // The structure of the FuncOp looks like the following:
+        // func.func @circuit(...) -> ... {...} {
+        // preprocessing
+        // quantum.device[...]
+        // tape 1
+        // quantum.device_release
+        // quantum.device[...]
+        // tape 2
+        // quantum.device_release
+        // ... more tapes
+        // post processing and return
+        // }
+
+        // This function parses the operations in the funcop and puts them into
+        // the container as
+        // OpsEachTape = [[preprocessing ops], [tape1 ops], [tape2 ops], ..., [postprocessing ops]]
+
+        for (size_t i = 0; i < OpsEachTape.size(); i++) {
+            OpsEachTape[i] = std::make_shared<SmallVector<Operation *>>();
+        }
+
+        // special: [0] for pre and [-1] for post processing
+        unsigned int cur_tape = 0;
+
+        func->walk([&](Operation *op) {
+            if (op == func) {
+                return; // don't visit the funcop itself
+            }
+            if (op->getName().getStringRef() == "quantum.device") {
+                cur_tape++;
+            }
+            OpsEachTape[cur_tape]->push_back(op);
+            if ((op->getName().getStringRef() == "quantum.device_release") &&
+                cur_tape == OpsEachTape.size() - 2) {
+                // reached post processing
+                cur_tape++;
+            }
+        });
+    } // collectOperationsForEachTape()
+
+    void collectNecessaryValuesFromEarlierTapes(const SmallVector<Operation *> &TapeOps,
+                                                SmallVector<Value> &NecessaryValuesFromEarlierTapes)
```

Comment:
Can you change this SmallVector to a set?

Reply:
Maybe same for tapeOps.

Reply:
Oh good idea!

Reply:
For some reason `SmallSet` does not work with `<Value>` but only `<Value *>`; however when I change all the corresponding places to `Value *` something breaks and the values are not handled appropriately...

Reply:
So I explored a bit here. `SmallSet` would make the set contain `Value *` instead of `Value`; however the in the function when looping through the operands of the operation, the `operand` variable (obtained from `getOperands()`) for each operand would all have the same address, causing issues. 

Not sure why this is happening.

Reply:
Note that `TapeOps` is meant to be ordered (it has insertion points set to its `front`, for example) and shouldn't be a set.

Reply:
Also the `NecessaryValuesFromEarlierTapes` is also ordered? Its order controls the return value's order of the outlined function:
```
        for (Value v : NecessaryValuesFromEarlierTapes) {
            Operation *VDefOp = v.getDefiningOp();
            if (std::find(TapeOps.begin(), TapeOps.end(), VDefOp) != TapeOps.end()) {
                // This Value needed for PP is in this tape!
                RetValues.push_back(v);
            }
        }
```

Reply:
> Why are you looking at its address?

In 
```
                for (auto operand : op->getOperands()) {
                    Operation *OperandSource = operand.getDefiningOp();
                    if (std::find(TapeOps.begin(), TapeOps.end(), OperandSource) == TapeOps.end()) {
                        // A list operand not produced in list itself, must be from earlier
                        // tapes/preprocessing!
                        if (std::find(NecessaryValuesFromEarlierTapes.begin(),
                                      NecessaryValuesFromEarlierTapes.end(),
                                      operand) == NecessaryValuesFromEarlierTapes.end()) {
                            NecessaryValuesFromEarlierTapes.push_back(operand);
                        }
                    }
                }
```

If `NecessaryValuesFromEarlierTapes` was to be a `SmallSet` instead of a `SmallVector`, then it needs to have `<Value *>` as contents, not `<Value>`, otherwise: 
```
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_function.h:400:20: error: invalid operands to binary expression ('const mlir::Value' and 'const mlir::Value')
      { return __x < __y; }
```
which is very strange because sets were meant to be unordered. A quick search on llvm repo reveals that `SmallSet<Value` is always using pointers: https://github.com/search?q=repo%3Allvm%2Fllvm-project%20smallset%3Cvalue&type=code

Reply:
> because the local stack variable can be reused across iterations to hold different objects

Ah, so that's why. I think there must be a way to extract the actual underlying `Value` for the operand. Let me explore a bit.

Reply:
> Ah, so that's why. I think there must be a way to extract the actual underlying Value for the operand. Let me explore a bit.

Yes, but I think you're going to run into the same problem, given that `Value` is a small object whose lifecycle is not normally bound to another context, hence the behaviour you are seeing regarding the stack address. This is different than an `Operation`, whose lifecycle is managed by the MLIR context I think, so using an `Operation *` makes sense as it points to a long-lived object on the heap.

> A quick search on llvm repo reveals that SmallSet<Value is always using pointers: https://github.com/search?q=repo%3Allvm%2Fllvm-project%20smallset%3Cvalue&type=code

This is interesting, although I'm not sure that an LLVM `Value` is the same type as an MLIR `Value` 🤔

Reply:
If hashability is the main problem, then an alternative would be to store a pair of owner and operand index `std::pair<Operation *, int>` in the set, which should allow you to reconstruct the `OpOperand`/`Value` object quite easily if needed.

Reply:
> Yes, but I think you're going to run into the same problem, given that `Value` is a small object whose lifecycle is not normally bound to another context, hence the behaviour you are seeing regarding the stack address. This is different than an `Operation`, whose lifecycle is managed by the MLIR context I think, so using an `Operation *` makes sense as it points to a long-lived object on the heap.

Interesting, never knew `Value` was short-lived. Let me try `<Operation*>` (I don't think there is a real need to stick to `Value`s in the pass)

Reply:
Small correction to what I said above, since "owner" is a bit ambiguous here. I think I meant owner (=operation who defined the value) + result index, which would allow you to reconstruct the `Value` quite easily (`owner->getResult(index)`).
The alternative owner (=operation who is _using_ the value) + operand index would allow you to reconstruct the `OpOperand` quite easily (`owner->getOperand(index)`), but then the same `Value` might appear multiple times in the set.

Reply:
One more thing I discovered about `Value`, it's a thin wrapper around `ValueImpl *` with no other members from what I can tell (the big reason why it's fine to pass around IR objects by value since they are very small):
`https://mlir.llvm.org/doxygen/Value_8h_source.html#l00096

I don't know where the impl is created, but it might be long-lived since its not in the `Value` class itself. In this case, you could probably identify each Value by its `ValueImpl *` in the set, although I don't know if that is good practice or not.

Reply:
I noticed the same pattern. I therefore propose the following:

- When finding necessary values from earlier tapes, go through the operands and record a `pair<Operation*, index>` to stand for the necessary value. The short-lived `Value` object does not go into the `NecessaryValuesFromEarlierTapes`. Instead we record the defining op (which is long-lived) and the index (which is just a number)
- When creating the return from the individual tapes, get that value from the `operation->getResult(index)`

This should resolve all ambiguities and facilitate in making `NecessaryValuesFromEarlierTapes` into a `SmallSet<std::pair<Operation*, unsigned int>>`

Reply:
> the following

done

Reply:
> you could probably identify each Value by its `ValueImpl *` in the set, although I don't know if that is good practice or not.

I don't think this is a standard thing to do. In mlir repository the only places using `ValueImpl` directly are places where `Value` (and some other core classes) are defined https://github.com/search?q=repo%3Allvm%2Fllvm-project%20path%3Amlir%20valueimpl&type=code, so I'd say it's not meant to be exposed for top-level uses

Reply:
You'd think that, but I have seen a lot `SmallVectorImpl` references in function signatures for some reason. Either way though I'm happy 👌

Reply:
Bookkeeping: this thread originally proposed changing both `NecesssaryValues...` and `TapeOps` from vectors to sets. 

`NecessaryValues..` is now done but this thread has become too long. I opened a new thread for `TapeOps`.